src/backends/r1cs/mod.rs

```diff
@@ -0,0 +1,557 @@
+        };
+
+        self.add_constraint(
+            "add constraint: x + cst = res",
+            Constraint{a, b, c},
+            span
+        );
+
+        res
+    }
+    
+    /// to constraint:
+    /// lhs * rhs = res
+    /// given:
+    /// a * b = c
+    /// then:
+    /// a = lhs
+    /// b = rhs
+    /// c = res
+    fn mul(&mut self, lhs: &CellVar, rhs: &CellVar, span: crate::constants::Span) -> CellVar {
+        let one = Fr::from(1);
+
+        let res = self.new_internal_var(Value::Mul(*lhs, *rhs), span);
+
+        let a = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(*lhs, one)])),
+            constant: None,
+        };
+
+        let b = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(*rhs, one)])),
+            constant: None,
+        };
+
+        let c = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(res, one)])),
+            constant: None,
+        };
+
+        self.add_constraint(
+            "mul constraint: lhs * rhs = res",
+            Constraint{a, b, c},
+            span
+        );
+
+        res
+    }
+    
+    /// to constraint:
+    /// x * cst = res
+    /// given:
+    /// a * b = c
+    /// then:
+    /// a = x
+    /// b = cst
+    /// c = res
+    fn mul_const(&mut self, x: &CellVar, cst: &Self::Field, span: crate::constants::Span) -> CellVar {
+        let one = Fr::from(1);
+
+        let res = self.new_internal_var(Value::Scale(*cst, *x), span);
+
+        let a = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(*x, one)])),
+            constant: None,
+        };
+
+        let b = LinearCombination {
+            terms: None,
+            constant: Some(*cst),
+        };
+
+        let c = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(res, one)])),
+            constant: None,
+        };
+
+        self.add_constraint(
+            "mul constraint: x * cst = res",
+            Constraint{a, b, c},
+            span
+        );
+
+        res
+    }
+    
+    /// to constraint:
+    /// x = cst
+    /// given:
+    /// a * b = c
+    /// then:
+    /// a = x
+    /// b = 1
+    /// c = cst
+    fn assert_eq_const(&mut self, x: &CellVar, cst: Self::Field, span: crate::constants::Span) {
+        let one = Fr::from(1);
+
+        let a = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(*x, one)])),
+            constant: None,
+        };
+
+        let b = LinearCombination {
+            terms: None,
+            constant: Some(one),
+        };
+
+        let c = LinearCombination {
+            terms: None,
+            constant: Some(cst),
+        };
+
+        self.add_constraint(
+            "eq constraint: x = cst",
+            Constraint{a, b, c},
+            span
+        );
+    }
+
+    /// to constraint:
+    /// lhs = rhs
+    /// given:
+    /// a * b = c
+    /// then:
+    /// a = lhs
+    /// b = 1
+    /// c = rhs
```

Comment:
also I think these kind of comments should be better as just comments in the implementation rather than in the rust doc comment, also you don't usually write rust doc comment on a trait implementation